subcommand/consul-sidecar/command.go

```diff
@@ -165,34 +166,44 @@ func (c *Command) Run(args []string) int {
 	// service hasn't changed and so won't update any indices. This means we
 	// won't be causing a lot of traffic within the cluster. We tolerate Consul
 	// Clients going down and will simply re-register once it's back up.
-	//
-	// The loop will only exit when the Pod is shut down and we receive a SIGINT.
-	for {
-		start := time.Now()
-		cmd := exec.CommandContext(signalCtx, c.flagConsulBinary, c.consulCommand...)
-
-		// Run the command and record the stdout and stderr output
-		output, err := cmd.CombinedOutput()
-		if err != nil {
-			c.logger.Error("failed to sync service", "output", strings.TrimSpace(string(output)), "err", err, "duration", time.Since(start))
-		} else {
-			c.logger.Info("successfully synced service", "output", strings.TrimSpace(string(output)), "duration", time.Since(start))
-		}
-		select {
-		// Re-loop after syncPeriod or exit if we receive interrupt or terminate signals.
-		case <-time.After(c.flagSyncPeriod):
-			continue
-		case <-signalCtx.Done():
-			// After the signal is received, wait for the merged metrics server
-			// to gracefully shutdown as well if it has been enabled. This can
-			// take up to metricsServerShutdownTimeout seconds.
-			if c.flagEnableMetricsMerging {
-				c.logger.Info("Attempting to shut down metrics server.")
-				c.shutdownMetricsServer(server)
+	go func() {
+		for {
+			start := time.Now()
+			cmd := exec.CommandContext(signalCtx, c.flagConsulBinary, c.consulCommand...)
+
+			// Run the command and record the stdout and stderr output
+			output, err := cmd.CombinedOutput()
+			if err != nil {
+				c.logger.Error("failed to sync service", "output", strings.TrimSpace(string(output)), "err", err, "duration", time.Since(start))
+			} else {
+				c.logger.Info("successfully synced service", "output", strings.TrimSpace(string(output)), "duration", time.Since(start))
+			}
+			select {
+			// Re-loop after syncPeriod or exit if we receive interrupt or terminate signals.
+			case <-time.After(c.flagSyncPeriod):
+				continue
+			case <-signalCtx.Done():
+				return
 			}
-			return 0
 		}
+	}()
+
+	// Block and wait for a signal or for the metrics server to exit.
+	select {
+	case <-signalCtx.Done():
+		// After the signal is received, wait for the merged metrics server
+		// to gracefully shutdown as well if it has been enabled. This can
+		// take up to metricsServerShutdownTimeout seconds.
+		if c.flagEnableMetricsMerging {
+			c.logger.Info("Attempting to shut down metrics server.")
+			c.shutdownMetricsServer(server)
+		}
+		return 0
+	case err := <-srvExitCh:
+		c.logger.Error(fmt.Sprintf("Metrics server error: %v", err))
+		return 1
 	}
+
 }
```

Comment:
below we discussed how `defer shutdownCancelFunc()` is not needed and how `server` cannot be `nil`.

Reply:
Ah sorry I had forgotten that. Added with a comment.

Reply:
Actually, the go vet check seems to be failing due to not calling the cancel func. I think we should still `defer shutdownCancelFunc()` and leave the comment about why its not necessary but required for go vet.